### PR TITLE
Linked Containers and Memory management

### PR DESCRIPTION
**Pull request by TeckniX**
Here's a quick PR for the linked container functionality and memory management. I'm sure both can be improved upon.

-- Signed: TeckniX <lokitek@gmail.com>

### CODE REVIEW

**Comment by Elesant** [comment]
Wow this is awesome! Thanks @TeckniX! Let me try it out locally.

**Comment by TeckniX** [comment]
I added some typeahead logic, but I noticed some bugs where a similar `pendingLink` (same with `pendingEnv`) is passed, then the view isn't being refreshed. 
I'm not sure why this is happening, and even forcing a reset of the state data with `replaceState` doesn't work. 

I'll let you guys decide how this should be handled. 

ps: I haven't updated the kitematic version since the recent addition, is this handled on a per release?

**Comment by jeffdm** [comment]
Taking a look, thanks @TeckniX . Will also help to merge this into the new flux refactoring

**Comment by TeckniX** [comment]
Forgot to reply, but I'm going through the flux tutorials to understand the difference and be able to implement the same changes to the flux refactoring. 
I'll make sure to do a PR once I get the flux part done.

**Comment by jeffdm** [comment]
@TeckniX awesome! Thanks for refactoring to flux so quickly :smiley: 

Taking a look!

File: src/components/ContainerSettingsLinks.react.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  },
+
+  getInitialState: function () {
+    let links = ContainerUtil.links(this.props.container) || [];
+    links.push(['', '']);
+    links = _.map(links, l => {
+      return [util.randomId(), l[0], l[1]];
+    });
+    let containers = containerStore.getState().containers;
+    let sorted = _.pluck(containers,'Name');
```
Comment (jeffdm) [inline src/components/ContainerSettingsLinks.react.js]: space after `,`

File: src/components/ContainerSettingsLinks.react.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+var ContainerSettingsLinks = React.createClass({
+  mixins: [React.addons.LinkedStateMixin],
+
+  contextTypes: {
+    router: React.PropTypes.func
+  },
+
+  getInitialState: function () {
+    let links = ContainerUtil.links(this.props.container) || [];
+    links.push(['', '']);
```
Comment (jeffdm) [inline src/components/ContainerSettingsLinks.react.js]: For a future PR:  maybe we can create a common 'list' component for env vars, links, basically any tables.

cc @Elesant

File: src/components/ContainerSettingsLinks.react.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    let vars = _.map(this.state.links, (kvp, index) => {
+      let [id, key, val] = kvp;
+      let icon;
+      if (index === this.state.links.length - 1) {
+        icon = <a onClick={this.handleAddLinksVar} className="only-icon btn btn-positive small"><span className="icon icon-add-1"></span></a>;
+      } else {
+        icon = <a onClick={this.handleRemoveLinksVar.bind(this, index)} className="only-icon btn btn-action small"><span className="icon icon-cross"></span></a>;
+      }
+
+      if(key == ""){
```
Comment (jeffdm) [inline src/components/ContainerSettingsLinks.react.js]: Space before and after ():

`if () {`

File: src/utils/ContainerUtil.js [hunk 4]
```diff
@@ -13,6 +13,48 @@ var ContainerUtil = {
...
+      if(key == "CpuShares" || key == "Memory" || key == "MemorySwap") {
+        SavedConfig[key] = val;
+      }
+    });
+    _.extend(SavedConfig, container.HostConfig);
+    if (!container || !container.HostConfig ) {
+      return [["CpuShares",""],["Memory",""],["MemorySwap",""]];
+    }
+    return _.map([
+      ['CpuShares',container.HostConfig.CpuShares],
```
Comment (jeffdm) [inline src/utils/ContainerUtil.js]: Space after `,`. Same below

File: src/utils/DockerUtil.js [hunk 5]
```diff
@@ -76,8 +76,45 @@ export default {
...
+          if(link.lastIndexOf('/') != -1 && link.lastIndexOf('/') > i) {
+            var valStart = link.lastIndexOf('/') + 1;
+          } else {
+            var valStart = i + 1;
+          }
+          var key = link.slice(keyStart, i);
+          var val = link.slice(valStart);
+
+          if(this.status(key)) {
+              startopts.Links.push(key+':'+val);
```
Comment (jeffdm) [inline src/utils/DockerUtil.js]: 2 space indent

File: src/utils/DockerUtil.js [hunk 6]
```diff
@@ -76,8 +76,45 @@ export default {
     } else {
       startopts.PublishAllPorts = true;
     }
+    // Add support for linked containers and runtime settings
+    if (containerData.HostConfig) {
+      var hostConfig = containerData.HostConfig;
+      if(typeof hostConfig.CpuShares !== "undefined" && hostConfig.CpuShares !== 0){
```
Comment (jeffdm) [inline src/utils/DockerUtil.js]: Replace `typeof hostConfig.CpuShares !== "undefined"` with `hostConfig.CpuShares` and similar below.

File: src/utils/DockerUtil.js [hunk 7]
```diff
@@ -274,6 +311,15 @@ export default {
     });
   },
 
+  status (name) {
+    let running = false;
+    let container = this.client.getContainer(name);
```
Comment (jeffdm) [inline src/utils/DockerUtil.js]: I don't think this actually returns a container, but rather a reference you can use to call `inspect`, `restart`, etc.

E.g.

```
let container = this.client.getContainer(name);
container.inspect(err, data => {});
```

It seems that this method would return false every time. Is it needed?

**Comment by jeffdm** [comment]
This is fantastic @TeckniX !

Some inline comments for code.

For UI, a few notes. But overall great implementation! Feels at home with the other settings panes :smiley: 

screen shot 2015-05-31 at 1 41 14 am



@mchiang0610 @Elesant Any feedback on the features + UI?

**Comment by TeckniX** [comment]
@jeffdm Thanks for all of the comments/notes - I'll go through them and start fixing the spacing issues and comment as needed

File: src/utils/DockerUtil.js (hunk 7 above)
Comment (TeckniX) [inline src/utils/DockerUtil.js]: This method was added to check the status of the container linked, preventing non-running (or incorrect) containers from being used for linking.

**Comment by jeffdm** [comment]
Thanks @TeckniX !

**Comment by mchiang0610** [comment]
@TeckniX 
The linking feature is super cool. 
- How does it specifically work? I was trying it out but I don't see it add a env variable to my other container. 

- Is there a specific use case for the Advanced section?

**Comment by TeckniX** [comment]
The linking should be just like the regular docker linking, so if you had a mysql DB container running and linked it to some nginx container, you would link it as such `--link mysql:mysql` which then shows up on inspection of the docker container. 

Your nginx env would then have something like:
```bash
MYSQL_PORT_3306_TCP_PORT=3306
MYSQL_PORT_3306_TCP=tcp://172.17.0.2:3306
MYSQL_PORT_3306_TCP_PROTO=tcp
MYSQL_NAME=/hungry_einstein/mysql
MYSQL_PORT_3306_TCP_ADDR=172.17.0.2
MYSQL_PORT=tcp://172.17.0.2:3306
```

**Comment by TeckniX** [comment]
The advance part of it was a request from : https://github.com/kitematic/kitematic/issues/258

**Comment by jeffdm** [comment]
@TeckniX Since linking is a critical feature, would you be okay only merging links in this PR? We can create a separate PR for the advanced piece. Linking will make combos like WP+Mysql, Drupal + MySQL so much easier for devs!

After a quick look I think #258 may refer to modifying the memory, cpu of the host VirtualBox VM, and not the containers themselves.

**Comment by TeckniX** [comment]
Totally fine with it, just trying to add some of the missing features as needed - Would you like me to split the PR ?

**Comment by TeckniX** [comment]
I'm going to close this thread and create a new PR

**Comment by RobLoach** [comment]
@TeckniX Do you know where the new PR for this went? I couldn't find it.

**Comment by FrenchBen** [comment]
Due to the new network ability of Docker, it was closed down.
See:
https://docs.docker.com/engine/userguide/containers/networkingcontainers/

**Comment by EpicVoyage** [comment]
Check out issue #1249 for the networking feature request.

**Comment by jhgorse** [comment]
@FrenchBen that link is ded. Perhaps this is the new one:
https://docs.docker.com/engine/tutorials/networkingcontainers/#network-containers

**Pull request closed without merging**